### PR TITLE
Docs: deployment example

### PR DESCRIPTION
**Pull request by xor-gate**
This is just a example how a borg backup server could be configured and used.
As this is just and idea, I keep it is open for discussion. Other people will also need to solve the centralised/offsite backup server problem.

### CODE REVIEW

File: docs/deployment.rst [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+
+The storage server needs to have only one UNIX user for all the clients.
+Recommended user and group and additional settings:
+
+* User: `backup`
+* Group: `backup`
+* Shell: e.g `/bin/bash` which is capable to run the `borg serve` command
+
+The clients probably will initiate a backup from the root user to catch all
+users, groups and permissions (e.g when backupping `/home`).
```
Comment (ThomasWaldmann) [inline docs/deployment.rst]: e.g. ?

File: docs/deployment.rst [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+* Repositories path (storage pool): /home/backup/repos
+* Clients restricted paths: `/home/backup/repos/<client fqdn>`
+  * johndoe.clnt.local: `/home/backup/repos/johndoe.clnt.local`
+  * web01.srv.local: `/home/backup/repos/web01.srv.local`
+  * app01.srv.local: `/home/backup/repos/app01.srv.local`
+
+Restrictions
+------------
+
+We need to restric the clients into their paths. Borg needs to be instructed to
```
Comment (ThomasWaldmann) [inline docs/deployment.rst]: restrict

File: docs/deployment.rst (hunk 1 above)
Comment (xor-gate) [inline docs/deployment.rst]: e.g: as an example
see: https://en.wiktionary.org/wiki/e.g.#Adverb

File: docs/deployment.rst [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+enforce this: ``borg serve --restrict-path /home/backup/repos/<client fqdn>``
+
+Clients can have one or more allowed ssh keys. Albeit recommended is to only use
+one per client. We add keys for ``johndoe.clnt.local``, ``web01.srv.local`` and
+``app01.srv.local``. Only they run under the user ``backup@backup01.srv.local``. A option
+per on-line of ``authorized_keys`` is added to distinguish between the clients:
```
Comment (ThomasWaldmann) [inline docs/deployment.rst]: the last sentence is strange, can you rephrase?

File: docs/deployment.rst [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+
+::
+
+  command="cd /home/backup/repos/<client fqdn>;
```
Comment (ThomasWaldmann) [inline docs/deployment.rst]: is the "cd" needed?

File: docs/deployment.rst [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+
+Client
+------
+
+The client needs to initialize the `pictures` repository like this:
+
+`borg init backup@backup01.srv.local:pictures`
+
+Or with the full path (should actually never be used, as only for demonstrational purposes).
+The server should automatic change the current working directory to the `<client fqdn>` folder.
```
Comment (ThomasWaldmann) [inline docs/deployment.rst]: ah, maybe because of this.

File: docs/deployment.rst (hunk 4 above)
Comment (xor-gate) [inline docs/deployment.rst]: This "puts the remote URL at the base of `pool/<client fqdn>`. And the user will not see any of the other clients by default.

File: docs/deployment.rst (hunk 5 above)
Comment (xor-gate) [inline docs/deployment.rst]: yes :_)

File: docs/deployment.rst (hunk 3 above)
Comment (xor-gate) [inline docs/deployment.rst]: Per ssh public key a option is added in the ``$HOME/.ssh/authorized_keys`` file to distinguish between the clients.

File: docs/deployment.rst [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+
+::
+
+  ~~~ SNIP ~~~
+  Remote: borg.remote.PathNotAllowed: /home/backup/repos/web01.srv.local/pictures
+  ~~~ SNIP ~~~
+  Repository path not allowed
+
+Ansible
+-------
```
Comment (ThomasWaldmann) [inline docs/deployment.rst]: I don't use ansible - is this useful in the situation you want to add another client later?

File: docs/deployment.rst [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+
+To automatic setup a storage server with the user, group, folders and
```
Comment (ThomasWaldmann) [inline docs/deployment.rst]: automate settting up

File: docs/deployment.rst [hunk 8]
```diff
@@ -0,0 +1,144 @@
...
+permissions a Ansible playbook could be used. Keep in mind the playbook
```
Comment (ThomasWaldmann) [inline docs/deployment.rst]: an

File: docs/deployment.rst (hunk 1 above)
Comment (en90) [inline docs/deployment.rst]: In context with the paragraph above, I think it's a strange way of phrasing. I would just say

>* Shell: `/bin/bash` (or other capable to run the `borg serve` command)

File: docs/deployment.rst [hunk 9]
```diff
@@ -0,0 +1,144 @@
...
+uses the Arch Linux `pacman<https://www.archlinux.org/pacman/pacman.8.html>`_
+package manager to install and keep borg up-to-date.
+
+Ansible takes care of all the system-specific commands to add the user, create the
+folder etcetera. And the configuration is human-readable.
```
Comment (ThomasWaldmann) [inline docs/deployment.rst]: etc. - and ...

File: docs/deployment.rst (hunk 6 above)
Comment (xor-gate) [inline docs/deployment.rst]: Yes because it automates the process over SSH from a management machine to satisfy the configuration.

File: docs/deployment.rst [hunk 10]
```diff
@@ -0,0 +1,144 @@
...
+      with_items: auth_users
+    - file: path="{{ home }}/.ssh/authorized_keys" owner="{{ user }}" group="{{ group }}" mode=0600 state=file
+    - file: path="{{ pool }}/{{ item.host }}" owner="{{ user }}" group="{{ group }}" mode=0700 state=directory
+      with_items: auth_users
+
+See also
+--------
+
+* `SSH Daemon manpage<http://www.openbsd.org/cgi-bin/man.cgi/OpenBSD-current/man8/sshd.8>`_
+* `Ansible<http://docs.ansible.com>`_
```
Comment (ThomasWaldmann) [inline docs/deployment.rst]: maybe this link is enough, you don't need to link to every piece of ansible docs you used.

Comment (xor-gate) [inline docs/deployment.rst]: agreed, as the ansible playbook is human readable by itself

File: docs/deployment.rst (hunk 1 above)
Comment (xor-gate) [inline docs/deployment.rst]: @en90 agreed

Comment (ThomasWaldmann) [inline docs/deployment.rst]: what i meant is to add the missing dot so it reads "e.g.".

File: docs/deployment.rst (hunk 4 above)
Comment (ThomasWaldmann) [inline docs/deployment.rst]: ok, good idea.

File: docs/deployment.rst (hunk 3 above)
Comment (ThomasWaldmann) [inline docs/deployment.rst]: Better. But: "an option"

Also, maybe use less passive voice (in general, maybe read through the whole thing to find/fix other places).
"We use per-ssh-key options in authorized_keys file to distinguish between the clients."

File: docs/deployment.rst [hunk 11]
```diff
@@ -0,0 +1,145 @@
...
+
+The storage server needs to have only one UNIX user for all the clients.
+Recommended user and group and additional settings:
+
+* User: `backup`
+* Group: `backup`
+* Shell: `/bin/bash` (or other capable to run the `borg serve` command)
+
+The clients probably will initiate a backup from the root user to catch all
+users, groups and permissions (e.g. when backupping `/home`).
```
Comment (ThomasWaldmann) [inline docs/deployment.rst]: can you integrate all the feedback already given?

File: docs/deployment.rst (hunk 6 above)
Comment (xor-gate) [inline docs/deployment.rst]: If you really want to you could reimplement this completely in "vanilla"-python or something else. But then it has to be copied and run locally on the storage server (as root user) which can be dangerous and break or kill the system.

**Comment by xor-gate** [comment]
I rebased all commits into one for easier merging.

File: docs/deployment.rst [hunk 12]
```diff
@@ -0,0 +1,146 @@
+.. include:: global.rst.inc
+.. _deployment:
+
+Deployment
+==========
+
+This chapter will give an example how to setup a borg storage server for multiple
```
Comment (ThomasWaldmann) [inline docs/deployment.rst]: maybe use "repository server" instead of "storage server"? (global replace)

File: docs/deployment.rst [hunk 13]
```diff
@@ -0,0 +1,146 @@
...
+  * app01.srv.local: `/home/backup/repos/app01.srv.local`
+
+Restrictions
+------------
+
+Borg is instructed to restrict clients into their own paths:
+``borg serve --restrict-path /home/backup/repos/<client fqdn>``
+
+Clients can have one or more allowed ssh keys. Albeit recommended is to only use
+one key per client. Keys are added for ``johndoe.clnt.local``, ``web01.srv.local`` and
```
Comment (ThomasWaldmann) [inline docs/deployment.rst]: instead of the first 2 sentences, you could just say "We use one ssh key per client here."

File: docs/deployment.rst [hunk 14]
```diff
@@ -0,0 +1,146 @@
...
+``backup@backup01.srv.local``. Every key in ``$HOME/.ssh/authorized_keys`` has a
+forced command and restrictions applied as shown below:
+
+::
+
+  command="cd /home/backup/repos/<client fqdn>;
+           borg serve --restrict-path /home/backup/repos/<client fqdn>",
+           no-port-forwarding,no-X11-forwarding,no-pty <keytype> <key> <host>
+
+**NOTE** The command above need to be compacted into a single line
```
Comment (ThomasWaldmann) [inline docs/deployment.rst]: The text shown above needs to be written on a single line.

File: docs/deployment.rst [hunk 15]
```diff
@@ -0,0 +1,146 @@
...
+
+The options which are added to the key will perform the following:
+
+1. Force ssh command: ``command="``
```
Comment (ThomasWaldmann) [inline docs/deployment.rst]: If you want to explain the above configuration, you actually should explain (and not repeat), like:

Force ssh command - when using this key, only this command is allowed.

File: docs/deployment.rst [hunk 16]
```diff
@@ -0,0 +1,146 @@
...
+2. Change working directory: ``cd /home/backup/repos/<client fqdn>``
```
Comment (ThomasWaldmann) [inline docs/deployment.rst]: don't repeat, but say WHY you do that, like:

Change working directory so the backup clients can use easy URLs like ssh://....

File: docs/deployment.rst [hunk 17]
```diff
@@ -0,0 +1,146 @@
...
+3. Run ``borg serve`` with restriction: ``borg serve --restrict-path /home/backup/repos/<client fqdn>"``
```
Comment (ThomasWaldmann) [inline docs/deployment.rst]: Run borg serve restricted to that client's base path.

File: docs/deployment.rst [hunk 18]
```diff
@@ -0,0 +1,146 @@
...
+4. Restrict ssh channel: ``no-port-forwarding,no-X11-forwarding,no-pty``
```
Comment (ThomasWaldmann) [inline docs/deployment.rst]: Restrict ssh and do not allow stuff we do not need.

File: docs/deployment.rst [hunk 19]
```diff
@@ -0,0 +1,146 @@
...
+
+Due to the cd command we use, the server automatically changes the current working
+directory so the client will not need to append the hostname to the remote URI.
```
Comment (ThomasWaldmann) [inline docs/deployment.rst]: if you explain this above, you can drop it here.

File: docs/deployment.rst (hunk 5 above)
Comment (ThomasWaldmann) [inline docs/deployment.rst]: still todo. in any case you need to fix grammar: automatic -> automatically.

File: docs/deployment.rst [hunk 20]
```diff
@@ -0,0 +1,146 @@
...
+The client needs to initialize the `pictures` repository like this:
+
+`borg init backup@backup01.srv.local:pictures`
+
+Or with the full path (should actually never be used, as only for demonstrational purposes).
+The server should automatic change the current working directory to the `<client fqdn>` folder.
+
+`borg init backup@backup01.srv.local:/home/backup/repos/johndoe.clnt.local/pictures`
+
+When `johndoe.clnt.local` tries to access a not restricted path the following error is raised.
```
Comment (ThomasWaldmann) [inline docs/deployment.rst]: "a not restricted path" is a bit misleading.
in fact you mean a disallowed path (which is any path outside of the directory borg serve is restricted to).

File: docs/deployment.rst (hunk 6 above)
Comment (ThomasWaldmann) [inline docs/deployment.rst]: no, rather keep ansible.

**Comment by xor-gate** [comment]
Rebased again ;_)

File: docs/deployment.rst [hunk 21]
```diff
@@ -0,0 +1,145 @@
...
+
+There is only one ssh key per client allowed. Keys are added for ``johndoe.clnt.local``, ``web01.srv.local`` and
+``app01.srv.local``. But they will access the backup under only one UNIX user account as:
+``backup@backup01.srv.local``. Every key in ``$HOME/.ssh/authorized_keys`` has a
+forced command and restrictions applied as shown below:
+
+::
+
+  command="cd /home/backup/repos/<client fqdn>;
+           borg serve --restrict-path /home/backup/repos/<client fqdn>",
```
Comment (ThomasWaldmann) [inline docs/deployment.rst]: i think borg usually adds some more options, e.g. --umask and --verbose or so, depending on what was given on the client's commandline. that doesn't work that flexibly and more if used with the forced server-side command, so the "relative path names" maybe should be solved otherwise. You could maybe just do it relative to the backup user's home directory.

Comment (xor-gate) [inline docs/deployment.rst]: If we really need to append the user commands (which is not very safe), then we need to wrap the `borg serve` command and fetch the necessary original ssh command arguments `$SSH_ORIGINAL_COMMAND`, see `man 1 ssh`. This example setup describes now a *forced* server which only restricts the path. The setup with passing arguments from the client to the server needs a more advanced script running on the server, maybe this should be described in a seperate chapter?

Comment (xor-gate) [inline docs/deployment.rst]: I had a look at:

## `--umask`

This doesn't make sense for the server side because all the backups are written under one user account.

## `--verbose`

Should only be enabled by the administrator per-ssh-key when there is something (very) wrong IMHO, because when running in verbose mode it could expose some info about the server.

Comment (ThomasWaldmann) [inline docs/deployment.rst]: well, wether --umask makes sense or not depends on what the default umask is and in some cases whether per-user groups is used or not. i introduce that option because i noticed that read permissions for "other" were given, which is undesirable for a backup storage (esp. if not encrypted). that's why it is now always used and defaults to a more safe "user-only" mask.

--verbose (on the server) should be used if the command on the client was invoked with it, otherwise you'll miss part of the expected output. I am not sure I can follow "it could expose info about the server".

Similar, it is for --lock-wait and --show-rc, they need to be given so that the server side behaves as expected.

Comment (xor-gate) [inline docs/deployment.rst]: About --umask you are right but if the parent path of the archives is set to 0700 this is no problem at all. 

I agree on passing from the client to the server the options but then I would need to write a command argument parser and filter out `--restrict-path` because else the client is able to pass a path himself. The wrapper needs then to fire-up the `borg serve --restrict-path <filtered args>` and then we should be good to go.

On the `--verbose` the server will print the linux version and distro (if we are on linux), and could be a security problem. But I thought when the client ssh-key is stolen they can remotely corrupt the backup anyway so the `--verbose` should be able to be passed.

Comment (ThomasWaldmann) [inline docs/deployment.rst]: about the wrapper: it could also do the "cd"

Comment (xor-gate) [inline docs/deployment.rst]: You mean we should cd into the "restricted path" and then directly add the `$SSH_ORIGINAL_COMMAND` ?

Comment (ThomasWaldmann) [inline docs/deployment.rst]: no, i mean that as the wrapper gets that directory anyway, it can do the "cd" as well as the processing you proposed.

Comment (xor-gate) [inline docs/deployment.rst]: I understand, but for the time being I will not extend this pull request with the wrapper script to allow client given parameters (as exercise to the reader) . The thing I could do is to add some `improvements` section and mention the improvement that needs to be made with some hinting.

Comment (xor-gate) [inline docs/deployment.rst]: As last note, maybe even `borg serve` could detect it is running under ssh by checking the `SSH_ORIGINAL_COMMAND` and cleanup/merge the arguments then there would not wrapper script needed at all. But this is totally out-of-scope of this pull request, there should be a issue/feature request opened if we want this.

Comment (ThomasWaldmann) [inline docs/deployment.rst]: ok, just add a note at that place that it currently ignores all options passed by the remote client.

**Comment by xor-gate** [comment]
I merged latest changed from master and squashed to only have one commit because the continues-integration failed I had not merged. The downside by squash is it rewrites all history and we can not see the small change I made.

**Comment by ThomasWaldmann** [comment]
ok, thanks!

**Comment by xor-gate** [comment]
@ThomasWaldmann, thanks for merging. I'm not a experienced RestructeredText writer and it seems some markup is broken when rendered by Readthedocs. At start of this document I tried to compile/generate html from the documentation but I got an error of something like: `python __version__.py` doesn't exist.

Have a look: http://borgbackup.readthedocs.org/en/latest/deployment.html

Specially the multi-level lists are broken in html, and the references on the bottom of the pages are broken.

**Comment by ThomasWaldmann** [comment]
Just have a look at the other rst files. You need an empty line before the list elements.

File: docs/deployment.rst [hunk 22]
```diff
@@ -0,0 +1,160 @@
...
+
+The options which are added to the key will perform the following:
+
+1. Force command on the ssh key and dont allow any other command to run
+2. Change working directory
+3. Run ``borg serve`` restricted at the client base path
+4. Restrict ssh and do not allow stuff which imposes a security risk
+
+Due to the cd command we use, the server automatically changes the current working
+directory so the client will not need to append the hostname to the remote URI.
```
Comment (ypid) [inline docs/deployment.rst]: "append the hostname" Shouldn’t it say "append the filepath"?

Comment (xor-gate) [inline docs/deployment.rst]: it is a bit misleading:

Due to the cd command we use, the server automatically changes the current working
directory **so the client will not need to append the hostname to the remote URI.**

Due to the `cd` command we use, the server automatically changes the current working
directory **then client doesn't need to have knowledge of the absolute or relative remote repository path**.

Comment (ypid) [inline docs/deployment.rst]: OK sure, so with hostname you meant `repos/<client fqdn>`? I can change that as I am currently going over the docs anyway and already have docs-fixes branch open :wink: Can I include your updated sentence?

Comment (xor-gate) [inline docs/deployment.rst]: @ypid, you may update this to the new sentense, if you find some other issues please correct them :+1: . Hostname is the lower part of the FQDN (fully qualified domain name, e.g.: `johndoe.mydomain.local` -> `johndoe`).

Comment (ypid) [inline docs/deployment.rst]: I included your change in https://github.com/borgbackup/borg/pull/56. Thanks.

Comment (xor-gate) [inline docs/deployment.rst]: @ypid, great
